src/ucs/profile/profile.c

```diff
@@ -226,143 +260,250 @@ static void ucs_profile_write()
-                                                ucs_profile_ctx.max_locations,
-                                                "profile_locations");
-        if (ucs_profile_ctx.locations == NULL) {
+    if (ucs_profile_global_ctx.num_locations > ucs_profile_global_ctx.max_locations) {
+        ucs_profile_global_ctx.max_locations =
+                        2 * ucs_profile_global_ctx.num_locations;
+        ucs_profile_global_ctx.locations =
+                        ucs_realloc(ucs_profile_global_ctx.locations,
+                                    sizeof(*ucs_profile_global_ctx.locations) *
+                                    ucs_profile_global_ctx.max_locations,
+                                    "profile_locations");
+        if (ucs_profile_global_ctx.locations == NULL) {
             ucs_warn("failed to expand locations array");
-            *loc_id_p = 0;
+            *loc_id_p = loc_id = 0;
             goto out_unlock;
         }
-
-        if (ucs_global_opts.profile_mode & UCS_BIT(UCS_PROFILE_MODE_ACCUM)) {
-            ucs_profile_ctx.accum.thread_locations =
-                            ucs_realloc(ucs_profile_ctx.accum.thread_locations,
-                                        sizeof(*ucs_profile_ctx.accum.thread_locations) *
-                                        ucs_profile_ctx.max_locations,
-                                        "profile_thread_locations");
-            if (ucs_profile_ctx.accum.thread_locations == NULL) {
-                ucs_warn("failed to expand thread locations array");
-                *loc_id_p = 0;
-                goto out_unlock;
-            }
-        }
     }
 
     /* Initialize new location */
-    loc             = &ucs_profile_ctx.locations[location];
+    loc = &ucs_profile_global_ctx.locations[ucs_profile_global_ctx.num_locations - 1];
     ucs_strncpy_zero(loc->super.file, basename(file), sizeof(loc->super.file));
     ucs_strncpy_zero(loc->super.function, function, sizeof(loc->super.function));
     ucs_strncpy_zero(loc->super.name, name, sizeof(loc->super.name));
     loc->super.line = line;
     loc->super.type = type;
     loc->loc_id_p   = loc_id_p;
 
-    if (ucs_global_opts.profile_mode & UCS_BIT(UCS_PROFILE_MODE_ACCUM)) {
-        ucs_profile_ctx.accum.thread_locations[location].total_time = 0;
-        ucs_profile_ctx.accum.thread_locations[location].count      = 0;
+out_found:
+    *loc_id_p = loc_id = (loc - ucs_profile_global_ctx.locations) + 1;
+    ucs_memory_cpu_store_fence();
+out_unlock:
+    pthread_mutex_unlock(&ucs_profile_global_ctx.mutex);
+    return loc_id;
+}
+
+static void ucs_profile_thread_expand_locations(int loc_id)
+{
+    ucs_profile_thread_context_t *ctx;
+    unsigned i, new_num_locations;
+
+    ctx = pthread_getspecific(ucs_profile_global_ctx.tls_key);
+    ucs_assert(ctx != NULL);
+
+    new_num_locations = ucs_max(loc_id, ctx->accum.num_locations);
+    ctx->accum.locations = ucs_realloc(ctx->accum.locations,
```

Comment:
it seems we don't need a new temporary variable to store `ucs_realloc()` result here as we call non-return `ucs_fatal` in case of error, right?

Reply:
yes, we don't need here